### PR TITLE
Migrate NPS code to gatekeeper

### PR DESCRIPTION
**Pull request by tejasbubane**
Along with token-based authentication. Per-company tokens are generated.

Original PR: https://github.com/Jifflenow/api_adapter/pull/174

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+git_source(:github) { |repo| "https://github.com/#{repo}.git" }
+
+ruby '2.5.5'
+
+# Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
+gem 'rails', '~> 5.2.3'
+# Use postgresql as the database for Active Record
+gem 'pg', '>= 0.18', '< 2.0'
+# Use Puma as the app server
+gem 'puma', '~> 3.11'
```
Comment (murali0861) [inline Gemfile]: not required

File: Gemfile [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+# Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
+gem 'jbuilder', '~> 2.5'
```
Comment (murali0861) [inline Gemfile]: not required

File: Gemfile [hunk 3]
```diff
@@ -0,0 +1,48 @@
+source 'https://rubygems.org'
+git_source(:github) { |repo| "https://github.com/#{repo}.git" }
+
+ruby '2.5.5'
```
Comment (murali0861) [inline Gemfile]: ruby will be picked up by respective machines please remove it.
we are using 2.3.5 in non development environments,

File: Gemfile [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+# Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
+gem 'rails', '~> 5.2.3'
```
Comment (murali0861) [inline Gemfile]: rails 5.2.1

File: README.md [hunk 5]
```diff
@@ -1,3 +1,13 @@
-# GateKeeper
+# README
 
-Common (hub level) `api_adapter` for all summit/portals.
+Common single installation `api_adapter` for all companies/events.
+
+* Ruby version - `2.5.5`
```
Comment (murali0861) [inline README.md]: update read me

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
+require 'common_utils/transporter_filter'
+
+class ApplicationController < ActionController::Base
+  include ApplicationHelper
+  include ErrorHandler::CustomResponseFormat
+  include CommonUtils::TransporterFilter
+
+  protect_from_forgery with: :null_session
```
Comment (murali0861) [inline app/controllers/application_controller.rb]: protect from forgery only for internal URL others please keep intact, please refer portal/api_adapter

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+  before_action :secured_internal_call?, :check_company, :authenticate_token,
+                :set_decryption_key, :decrypt_transporter_payload
+
+  INTERNAL_MIME = 'internal'
+
+  def error_handler(response, re)
+    Rails.logger.warn re.inspect
+    response[:status] = APIConstants::Status_Invalid_Params
```
Comment (murali0861) [inline app/controllers/application_controller.rb]: response is default instance variable in rails please rename it

File: app/controllers/configurations_controller.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
+class ConfigurationsController < ApplicationController
+  skip_before_action :check_company, :authenticate_token,
+                     :set_decryption_key, :decrypt_transporter_payload
+  before_action :check_for_basic_auth
+
+  def create
+    response = initialize_response_hash
```
Comment (murali0861) [inline app/controllers/configurations_controller.rb]: response is default instance variable in rails please rename it

File: app/controllers/nps_controller.rb [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+      method: :post,
+      body: build_params,
+      headers: {'Content-Type'=> "application/x-www-form-urlencoded"},
+      connecttimeout: 10
+    }
+    begin
+      CustomLogger.info "Call to be create SFDC support ticket:: "\
+                        "URL: #{uri}\nData: #{build_params.inspect}"
+      Typhoeus::Request.new(uri.to_s, typheous_options).run
+    rescue => re
```
Comment (murali0861) [inline app/controllers/nps_controller.rb]: rescue from runtime error only and newrelic notify when error happened

File: app/mailers/application_mailer.rb [hunk 10]
```diff
@@ -0,0 +1,4 @@
+class ApplicationMailer < ActionMailer::Base
+  default from: 'from@example.com'
```
Comment (murali0861) [inline app/mailers/application_mailer.rb]: for NPS u need to use the summit/portal default email configured.

File: config/database.yml [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+  <<: *default
+  database: gatekeeper_development
+
+test:
+  <<: *default
+  database: gatekeeper_test
+
+production:
+  <<: *default
+  database: gatekeeper_prd
```
Comment (murali0861) [inline config/database.yml]: don't add for prod

File: config/puma.rb [hunk 12]
```diff
@@ -0,0 +1,34 @@
+# Puma can serve each request in a thread from an internal thread pool.
```
Comment (murali0861) [inline config/puma.rb]: not required, at this moment we are not using puma

File: config/storage.yml [hunk 13]
```diff
@@ -0,0 +1,34 @@
+test:
```
Comment (murali0861) [inline config/storage.yml]: i don't think we are using any kind of storage

File: db/migrate/20190715055106_create_company_configurations.rb [hunk 14]
```diff
@@ -0,0 +1,10 @@
+class CreateCompanyConfigurations < ActiveRecord::Migration[5.2]
+  def change
+    create_table :company_configurations do |t|
+      t.string :company
+      t.json :config
```
Comment (murali0861) [inline db/migrate/20190715055106_create_company_configurations.rb]: use jsonb

File: app/mailers/application_mailer.rb (hunk 10 above)
Comment (tejasbubane) [reply app/mailers/application_mailer.rb]: My bad, we are not using mailer here. It is in that `feedback_handler` gem which will use summit/portal's application_mailer. I have removed this mailer.

File: config/database.yml (hunk 11 above)
Comment (tejasbubane) [reply config/database.yml]: Added `database.yml` to gitignore and created `database_example.yml`.

File: db/migrate/20190715055106_create_company_configurations.rb (hunk 14 above)
Comment (tejasbubane) [reply db/migrate/20190715055106_create_company_configurations.rb]: changed to `jsonb`.

**Comment by tejasbubane** [comment]
Closing in favour of #2. Renamed branch to `nps-gatekeeper` to be in sync with other repos.

**Pull request closed without merging**